src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -3363,6 +3363,22 @@ private void SerializeMethodSpecificationSignature(BlobBuilder builder, IGeneric
             }
         }
 
+        private EmitContext GetEmitContextForAttribute(ICustomAttribute customAttribute)
+        {
+            if (customAttribute is AttributeData attributeData &&
+                attributeData.ApplicationSyntaxReference?.GetSyntax(_cancellationToken) is { } syntaxNode)
```

Comment:
>attributeData.ApplicationSyntaxReference?.GetSyntax(_cancellationToken) is { } syntaxNode [](http://example.com/codeflow?start=16&length=89)

> I wasn't sure which approach is better. But getting the syntax node is probably an expensive operation, so I guess you're right. 

I was mostly concerned about amount of ```EmitContext``` instances that we were allocating. However, your comment got me thinking about a different aspect. Getting syntax from a syntax reference can actually resurrect unloaded syntax trees. And we will be doing this for every tree that has attributes. This might have a negative performance impact in IDE scenarios. I think we should switch ```EmitContext``` to hold on to syntax reference instead, and request the actual syntax only when we actually need it (reporting an error, etc.) This refactoring should be performed separately, let's open a follow up issue for that.

Reply:
#66371